### PR TITLE
RSCBC-62: Refactored hostname handling when parsing configs

### PR DESCRIPTION
**Pull request by chvck**
Motivation
----------
We're currently replacing $HOST in configs with the endpoint reference rather than the source hostname.

Changes
-------
Refactors how we deal with remote hostnames and addresses from the memdx package, to lift addresses in errors up to the core level.
Fix SRV record resolving and connecting to FQDNs.

### CODE REVIEW

File: sdk/couchbase-core/src/agent.rs [hunk 1]
```diff
@@ -528,13 +530,12 @@ impl Agent {
     fn gen_first_kv_client_configs(
         memd_addrs: &Vec<String>,
         state: &AgentState,
-    ) -> HashMap<String, KvClientConfig> {
+    ) -> Result<HashMap<String, KvClientConfig>> {
```
Comment (Peter-Searby) [inline sdk/couchbase-core/src/agent.rs:533]: Why does this need to be a Result? It doesn't look like any error case was added?

File: sdk/couchbase-connstr/src/lib.rs [hunk 2]
```diff
@@ -50,7 +52,7 @@ impl ConnSpec {
             let scheme = scheme_type.as_str();
             if (scheme != "couchbase" && scheme != "couchbases")
                 || self.hosts.len() != 1
-                || self.hosts[0].port.is_none()
+                || self.hosts[0].port.is_some()
```
[the remark this answers is not in the document]
Comment (chvck) [reply sdk/couchbase-connstr/src/lib.rs:55]: This just fixes the SRV record resolving. Changing the address handling all the way down into memdx fixes the FQDNs - SocketAddr cannot hold a FQDN - which makes sense really...

File: sdk/couchbase-core/src/agent.rs (hunk 1 above)
Comment (chvck) [reply sdk/couchbase-core/src/agent.rs]: Good spot, a carry-over from when I was trying to fix the FQDN bits.

File: sdk/couchbase-connstr/Cargo.toml [hunk 3]
```diff
@@ -6,6 +6,10 @@ edition = "2021"
 [dependencies]
 form_urlencoded = "1.2.1"
 hickory-resolver = { version = "0.24.1", features = ["dns-over-rustls"], default-features = false }
+log = "0.4.22"
 regex = "1.11.0"
 thiserror = "1.0"
 url = "2.5.2"
+
+[dev-dependencies]
+tokio = { version = "1.0", features = ["rt", "net", "macros"] }
```
Comment (Peter-Searby) [inline sdk/couchbase-connstr/Cargo.toml:15]: This looks like a very old version of tokio. We seem to have "1.40" and "1.38" in the existing cargo.toml files. I'm not too well informed on the rust dependency management best practices, but this doesn't seem great?

File: sdk/couchbase-core/src/agent.rs [hunk 4]
```diff
@@ -528,13 +530,12 @@ impl Agent {
     fn gen_first_kv_client_configs(
         memd_addrs: &Vec<String>,
         state: &AgentState,
-    ) -> HashMap<String, KvClientConfig> {
+    ) ->HashMap<String, KvClientConfig> {
```
Comment (Peter-Searby) [inline sdk/couchbase-core/src/agent.rs:533]: Nit: Looks like some spacing was lost.
I'm honestly quite surprised there's not something like rustfmt being used in this codebase

File: sdk/couchbase-core/src/configwatcher.rs [hunk 5]
```diff
@@ -144,18 +144,21 @@ where
...
-        let addr = client.remote_addr();
-        let host = addr.ip();
+        let hostname = client.remote_hostname();
 
         let resp = client
             .get_cluster_config(GetClusterConfigRequest {})
             .await?;
 
-        let config: TerseConfig = serde_json::from_slice(resp.config.as_slice())?;
+        let config = parse_terse_config(&resp.config, hostname)?;
```
Comment (Peter-Searby) [inline sdk/couchbase-core/src/configwatcher.rs:157]: This is a bit confusingly named, considering the call of another function of the same name (albeit a different namespace) immediately after

File: sdk/couchbase-core/src/agent.rs (hunk 4 above)
Comment (chvck) [reply sdk/couchbase-core/src/agent.rs:533]: I thought we had it running rustfmt checking on PR, will file an issue to add.

File: sdk/couchbase-connstr/Cargo.toml (hunk 3 above)
Comment (chvck) [reply sdk/couchbase-connstr/Cargo.toml]: Not entirely sure why that was 1.0, it's now 1.40.

File: sdk/couchbase-core/src/configwatcher.rs (hunk 5 above)
Comment (chvck) [reply sdk/couchbase-core/src/configwatcher.rs]: I've added the module name to the call to try to make it more clear. IMO it is the correct name with regards to the module itself though.

**Review by Matt-Woz — APPROVED** [review]

**Pull request merged**